src/ServerRequestFactory.php

```diff
@@ -99,6 +101,82 @@ public static function fromGlobals(
+        }
+        $get    = isset($request->get) ? $request->get : [];
+        $post   = isset($request->post) ? $request->post : [];
+        $cookie = isset($request->cookie) ? $request->cookie : [];
+        $files  = isset($request->files) ? $request->files : [];
+
+        $server = [
+            'REQUEST_METHOD' => $request->server['request_method'],
+            'REQUEST_URI' => $request->server['request_uri'],
+            'PATH_INFO' => $request->server['path_info'],
+            'REQUEST_TIME' => $request->server['request_time'],
+            'GATEWAY_INTERFACE' => 'swoole/' . SWOOLE_VERSION,
+            // Server
+            'SERVER_PROTOCOL' => isset($request->header['server_protocol']) ?
+                                 $request->header['server_protocol'] : $request->server['server_protocol'],
+            'REQUEST_SCHEMA' => isset($request->header['request_scheme']) ?
+                                $request->header['request_scheme'] :
+                                explode('/', $request->server['server_protocol'])[0],
+            'SERVER_NAME' => isset($request->header['server_name']) ?
+                             $request->header['server_name'] : $request->header['host'],
+            'SERVER_ADDR' => $request->header['host'],
+            'SERVER_PORT' => isset($request->header['server_port']) ?
+                             $request->header['server_port'] : $request->server['server_port'],
+            'REMOTE_ADDR' => isset($request->server['remote_addr']) ?
+                             $request->server['remote_addr'] : $request->header['host'],
+            'REMOTE_PORT' => isset($request->header['remote_port']) ?
+                             $request->header['remote_port'] : $request->server['remote_port'],
+            'QUERY_STRING' => isset($request->server['query_string']) ?
+                              $request->server['query_string'] : '',
+            // Headers
+            'HTTP_HOST' => $request->header['host'],
+            'HTTP_USER_AGENT' => isset($request->header['user-agent']) ?
+                                 $request->header['user-agent'] : '',
+            'HTTP_ACCEPT' => isset($request->header['accept']) ?
+                             $request->header['accept'] : '*/*',
+            'HTTP_ACCEPT_LANGUAGE' => isset($request->header['accept-language']) ?
+                                      $request->header['accept-language'] : '',
+            'HTTP_ACCEPT_ENCODING' => isset($request->header['accept-encoding']) ?
+                                      $request->header['accept-encoding'] : '',
+            'HTTP_CONNECTION' => isset($request->header['connection']) ?
+                                 $request->header['connection'] : '',
+            'HTTP_CACHE_CONTROL' => isset($request->header['cache-control']) ?
+                                    $request->header['cache-control'] : '',
+        ];
```

Comment:
I'd argue we shouldn't populate most of the above under Swoole; most of these are things that an SAPI populates _because there is no other way to get at headers, the request URI parts, etc._. Since we _do_ have a way to get at those, I'd argue we should only populate the `SERVER_*` and `REMOTE_*` values, and maybe the `GATEWAY_INTERFACE` and `REQUEST_TIME`.

Also, please align these as follows:

```php
'key' => isset(...)
    ? ...
    : ...,
```

When it comes to marshaling the URI, we have the various parts in the Swoole request, and can pass a smaller array _in situ_ when calling `marshalUriFromServer()`:

```php
static::marshalUriFromServer([ /* values we need here */ ], $headers)
```

Reply:
This is the `var_dump` of a Swoole HTTP request, using this simple [test example](https://gist.github.com/ezimuel/c2613f771c415678dc50d1d1930585cd).

```
object(Swoole\Http\Request)#7 (9) {
  ["fd"]=>
  int(1)
  ["header"]=>
  array(5) {
    ["host"]=>
    string(14) "localhost:9501"
    ["user-agent"]=>
    string(12) "HTTPie/0.9.8"
    ["accept-encoding"]=>
    string(13) "gzip, deflate"
    ["accept"]=>
    string(3) "*/*"
    ["connection"]=>
    string(10) "keep-alive"
  }
  ["server"]=>
  array(11) {
    ["request_method"]=>
    string(3) "GET"
    ["request_uri"]=>
    string(1) "/"
    ["path_info"]=>
    string(1) "/"
    ["request_time"]=>
    int(1527775742)
    ["request_time_float"]=>
    float(1527775742.4968)
    ["server_port"]=>
    int(9501)
    ["remote_port"]=>
    int(32922)
    ["remote_addr"]=>
    string(9) "127.0.0.1"
    ["master_time"]=>
    int(1527775742)
    ["server_protocol"]=>
    string(8) "HTTP/1.1"
    ["server_software"]=>
    string(18) "swoole-http-server"
  }
  ["request"]=>
  NULL
  ["cookie"]=>
  NULL
  ["get"]=>
  NULL
  ["files"]=>
  NULL
  ["post"]=>
  NULL
  ["tmpfiles"]=>
  NULL
}
```
I think we should include all these values, what do you think?